admin/server/auth/pkce.go

```diff
@@ -0,0 +1,178 @@
+package auth
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/rilldata/rill/cli/pkg/auth"
+)
+
+const authorizationCodeGrantType = "authorization_code"
+
+// in-memory storage for authorization codes
+// TODO use persistent storage to handle admin server restarts, but since codes expire in a minute, is it really necessary, user can just re-initiate login ?
+var authCodeDB = make(map[string]AuthorizationCode)
+
+// AuthorizationCode represents the stored information for an authorization code
+type AuthorizationCode struct {
+	ClientID            string
+	RedirectURI         string
+	UserID              string
+	Expiration          time.Time
+	CodeChallenge       string
+	CodeChallengeMethod string
+}
+
+func handlePKCE(w http.ResponseWriter, r *http.Request, clientID, userID, codeChallenge, codeChallengeMethod, redirectURI string) {
+	// Generate a unique authorization code
+	code, err := generateRandomString(16) // 16 bytes, resulting in a 32-character hex string
+	if err != nil {
+		http.Error(w, "Failed to generate authorization code", http.StatusInternalServerError)
+		return
+	}
+
+	// Set the expiration date for the authorization code (e.g., a minute from now)
+	// Note from https://www.oauth.com/oauth2-servers/authorization/the-authorization-response/
+	// The authorization code must expire shortly after it is issued. The OAuth 2.0 spec recommends a maximum lifetime of 10 minutes, but in practice, most services set the expiration much shorter, around 30-60 seconds.
+	expiration := time.Now().Add(1 * time.Minute)
+
+	authCodeDB[code] = AuthorizationCode{
+		ClientID:            clientID,
+		RedirectURI:         redirectURI,
+		UserID:              userID,
+		Expiration:          expiration,
+		CodeChallenge:       codeChallenge,
+		CodeChallengeMethod: codeChallengeMethod,
+	}
+
+	// Build the redirection URI with the authorization code as per OAuth2 spec
+	redirectWithCode := fmt.Sprintf("%s?code=%s&state=%s", redirectURI, code, r.URL.Query().Get("state"))
```

Comment:
Could there ever be non-URL safe characters in the state? If yes, then consider using the `url` package to parse, add to and reconstruct the url

Reply:
Okay, can we also add a small comment here to note that both the state and code are URL safe then? Just to save the next reader the time to trace how they are generated